projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs

```diff
@@ -262,51 +267,105 @@ public void SendHeader()
 
         public void Write(OutgoingFrame frame)
         {
-            _channelWriter.TryWrite(frame);
+            LockToken token = default;
+            try
+            {
+                token = _singleWriterMutex.TryWait(WaitOptions.NoDelay);
+                if (!token.Success)
+                {
+                    // Didn't get the lock immediately, let's backlog the write and start the backlog writer task.
```

Comment:
General question, this makes it possible that the order of published messages could get changed. Is this "allowed"? Does any documentation say something about this?

Reply:
Can they? Considering that we are using a write lock, which only one can hold at a time, if we can't get the lock synchronously, we'll add the write to the backlog, since the backlog is the one holding the lock. I'll take a closer look at this as well, I'm shamelessly stealing this logic from StackExchange.Redis :)

Reply:
Good question,  my thought process was:
- T1 has the lock, WriteFrameToPipe
- T2 didn't get the lock, wrote frame to channel, about to start the Task.Run
- T3 just entered the method

=> If T1 releases the lock & T3 gets the lock before the channel gets it, it will publish T3 before T2's message.

But now that I think of it, this is most probably irrelevant, as it uses multiple threads which the order will never be guaranteed anyway...